Mimic xenserver

ydirson opened the pull request:
stormi: Bring back file layout from previous versions of XCP-ng / XenServer, some of which is for very good reasons, the other being perfectible or possibly even useless sometimes but works.

We'll still drop some files which we really can't find a use for in the boot process: grub-usb.cfg, and duplicates of the grub bootloader in /boot.

ydirson committed:
create-install-iso: generate efiboot.img's grub.cfg from standard one

Generating it will avoid desynchronisation experienced in the past.

There seem to be no point in keeping a copy named "grub-usb.cfg" like XS
does, inside the ISO where GRUB will never look for it, its role is only
for use inside efiboot.img.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
README: show how to testboot from USB-storage with qemu

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
create-install-iso: files for PXE boot

All in a new centralized /pxe-files directory, and in various places for
compatibility with XCP-ng 8.2 and XenServer.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

stormi commented on scripts/create-install-iso.sh:
```diff
@@ -287,6 +287,8 @@ else
     "${FAKETIME[@]}" mformat -i "$ISODIR/boot/efiboot.img" -N 0 -C -f 2880 -L 16 ::.
     "${FAKETIME[@]}" mmd     -i "$ISODIR/boot/efiboot.img" ::/EFI ::/EFI/BOOT
     "${FAKETIME[@]}" mcopy   -i "$ISODIR/boot/efiboot.img" "$BOOTX64" ::/EFI/BOOT/BOOTX64.EFI
+    "${FAKETIME[@]}" mcopy   -i "$ISODIR/boot/efiboot.img" "$ISODIR/EFI/xenserver/grub-usb.cfg" ::/EFI/BOOT/GRUB.CFG
```
To completely mimic XS, the target name should be `EFI/xenserver/grub.cfg`. That's where the grub binary from the RPM would look for, in hypothetical situations where it would look inside the filesystem embedded into efiboot.img, instead of the ISO's main filesystem.

ydirson replied:
oops

ydirson committed:
README: fix consistency of example commands

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
create-install-iso: capitalization inside efiboot.img

This is for extra safety only, we don't even know what mtools does when
creating an 8.3 entry with upper or lower case, there are always uppercase
letters in the fs, and only mtools seems to be able (how!?) to remember
the case it created them with.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
create-install-iso: put grub files in /EFI/xenserver not /boot/grub

Using a non-default path allows to have a dedicated subtree on a PXE
server.

/EFI/xenserver/ is used for layout compatibility with
XenServer-derived GRUB packages (see next commit for `--efi-installer
rpm`).

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
create-install-iso: regroup code related to efiboot.img building

We want to make this code optional to be able to use a prebuilt version
itself, this avoids to reorder and reindent lines in a single commit.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
create-install-iso: by default use grubx64.efi provided by grub-efi package

We keep the ability to call mkimage, through the new --efi-installer flag.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
create-install-iso: include EULA and LICENSE in ISO root

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
create-install-iso: update some comments for better clarity

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
Generate .treeinfo from branding

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

stormi commented on scripts/create-install-iso.sh:
```diff
@@ -263,25 +263,27 @@ if false; then
-    "${FAKETIME[@]}" mcopy   -i "$ISODIR/boot/efiboot.img" "$BOOTX64" ::/EFI/BOOT/BOOTX64.EFI
 
     # grub modules
     # FIXME: too many modules?
```
It's not modules anymore

ydirson replied:
that part is still in the branch taking the modules

stormi commented on scripts/create-install-iso.sh:
```diff
@@ -268,18 +268,22 @@ else
-    # FIXME: too many modules?
-    tar -C "$SCRATCHDIR/grub/usr/lib" -cf - grub/x86_64-efi |
-        tar -C "$ISODIR/boot" -xf - ${VERBOSE}
+    if false; then
```
Either we drop it, or add a comment explaining what this skipped part of the script use to do?

ydirson replied:
will switch to alternate operating modes selectable from CLI

stormi commented on templates/iso/8.3/EFI/xenserver/grub-usb.cfg:
```diff
@@ -53,3 +53,6 @@ menuentry "shell" {
 }
 
+menuentry "memtest" {
+    chainloader /boot/memtest.efi
```
I wonder if this would work in all boot scenarios (especially when/if the contents of efiboot.img becomes the root filesystem for EFI boot).

ydirson replied:
I fear we're only going to find out when someone with that kind of BIOS tries. We're likely to find out earlier, if we don't put grub.img at all inside `efiboot.img`, as then we'll be able to gather real information and get someone to test all cases.

ydirson commented:
* added the ".treeinfo from branding" support
* flagged WIP the 2 commits we still need to focus on
